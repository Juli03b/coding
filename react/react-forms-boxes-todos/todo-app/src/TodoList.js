import { useState } from "react";
import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";
import "./TodoList.css"

const TodoList = () => {
    const [todos, setTodos] = useState([]);
    const [editMode, setEditMode] = useState(false);
    const [editData, setEditData] = useState("");
    const [todoToEdit, setTodoToEdit] = useState();
    const addTodo = (todo) => {
        setTodos( todos => [...todos, todo]);
    }
    const removeTodo = (idx) => {
        setTodos( todos => todos.filter( (todo, j) => j !== idx));
    }
    const handleEditEnable = (idx) => {
        setEditMode( (editMode) => !editMode);
        setTodoToEdit(idx);
    }
    const handleEditChange = (evt) => {
        const { value } = evt.target;
        setEditData( data => value);
    }
    const handleEditSubmit = (evt) => {
        evt.preventDefault();
        setTodos( todos => 
            todos.map( (todo, idx) => {
                if(idx === todoToEdit){
                    return editData;
                }
                return todo;
            })
        )
    }

    return (
        <>
            <NewTodoForm addTodo={addTodo} />
            <ul>
                {todos.map( (todo, idx) => <Todo todo={todo} removeTodo={() => removeTodo(idx)} editTodo={() => handleEditEnable(idx) } />)}
            </ul>
            { editMode && 
                <form onSubmit={handleEditSubmit} data-testid="todo-edit-form" >
                    <input value={editData} placeholder="Edit todo" onChange={handleEditChange} data-testid="todo-edit-input" />
                    <button>Update Todo</button>
                </form>
            }
        </>
    );
}

export default TodoList;