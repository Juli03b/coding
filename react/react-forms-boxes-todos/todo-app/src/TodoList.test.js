import { fireEvent, render } from "@testing-library/react";
import TodoList from "./TodoList";

it("Should render", () => {
    render(<TodoList />);    
});

it("Should match snapshots", () => {
    const { asFragment } = render(<TodoList />);    

    expect(asFragment()).toMatchSnapshot();
});

describe('Test typing a task, submitting, and editing', () => {
    const { getByTestId, getAllByTestId } = render(<TodoList />);    
    const newTodoForm = getByTestId("new-todo-form");
    const newTodoInput = getByTestId("new-todo-input");
    
    fireEvent.change(newTodoInput, {target: {value: "Go outside"}});
    fireEvent.submit(newTodoForm);

    const todo = getByTestId("todo");
    expect(todo).toBeInTheDocument();

    fireEvent.click(todo);
    
    const editForm = getByTestId("todo-edit-form");
    const editInput = getByTestId("todo-edit-input");
    expect(editForm).toBeInTheDocument();
    expect(editInput).toBeInTheDocument();

    fireEvent.change(editInput, {target: {value: "EAT"}});
    fireEvent.submit(editForm);


    const todoUpdated = getByTestId("todo");
    expect(todoUpdated).toBeInTheDocument();
});
