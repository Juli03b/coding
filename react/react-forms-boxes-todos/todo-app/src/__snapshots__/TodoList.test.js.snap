// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshots 1`] = `
<DocumentFragment>
  <form
    data-testid="new-todo-form"
  >
    <label
      for="todo"
    >
      Todo:
    </label>
    <input
      data-testid="new-todo-input"
      id="todo"
      name="todo"
      placeholder="Todo"
      value=""
    />
    <button>
      Submit todo!
    </button>
  </form>
  <ul />
</DocumentFragment>
`;
